File: typescript/src/models/Location.ts
/**
 * Location model
 * Converted from go/models/location.go
 * Owner: location-navigation-service
 * Consumers: backend, dispatch-asset-service, mission-command-service
 */

import { Coordinate } from './common';

/** Area represents a named zone within a location with polygon boundary */
export interface Area {
  id: string; // UUID within location
  name: string;
  description?: string;
  boundary: Coordinate[]; // Polygon points
  fillColor?: string;
  borderColor?: string;
  opacity?: number; // 0.0 to 1.0
  type?: string; // perimeter, patrol_zone, checkpoint, etc.
  priority?: string; // high, medium, low
  active: boolean;
  createdAt: string; // ISO 8601
  updatedAt: string; // ISO 8601
}

/** Location represents a geographic location with center coordinates and multiple areas */
export interface Location {
  id: string;
  name: string;
  description?: string;

  // Center coordinates
  latitude: number;
  longitude: number;

  // Areas within this location
  areas: Area[];

  // Visual settings
  color?: string; // Hex color for map display
  icon?: string; // Icon identifier

  // Metadata
  useCase?: string; // military, police, defense, etc.
  tags?: string[];

  // Status
  active: boolean;

  // Audit
  createdBy?: string;
  createdAt: string; // ISO 8601
  updatedBy?: string;
  updatedAt: string; // ISO 8601
}

/** CreateLocationRequest represents the request to create a new location */
export interface CreateLocationRequest {
  name: string;
  description?: string;
  latitude: number;
  longitude: number;
  color?: string;
  icon?: string;
  useCase?: string;
  tags?: string[];
  active?: boolean;
}

/** UpdateLocationRequest represents the request to update a location */
export interface UpdateLocationRequest {
  name?: string;
  description?: string;
  latitude?: number;
  longitude?: number;
  color?: string;
  icon?: string;
  useCase?: string;
  tags?: string[];
  active?: boolean;
}

/** CreateAreaRequest represents the request to create a new area */
export interface CreateAreaRequest {
  name: string;
  description?: string;
  boundary: Coordinate[]; // min 3 points
  fillColor?: string;
  borderColor?: string;
  opacity?: number;
  type?: string;
  priority?: string;
  active?: boolean;
}

/** UpdateAreaRequest represents the request to update an area */
export interface UpdateAreaRequest {
  name?: string;
  description?: string;
  boundary?: Coordinate[];
  fillColor?: string;
  borderColor?: string;
  opacity?: number;
  type?: string;
  priority?: string;
  active?: boolean;
}

File: python/src/phlx_contracts/models/location.py
"""Location model for Phylax platform"""

from typing import Optional, List
from datetime import datetime
from enum import Enum
from pydantic import BaseModel, Field
from .common import Coordinate

# Owner: location-navigation-service
# Consumers: backend, dispatch-asset-service, mission-command-service


class LocType(str, Enum):
    """Location type enumeration"""

    PERIMETER = "perimeter"
    PATROL_ZONE = "patrol_zone"
    CHECKPOINT = "checkpoint"


class Area(BaseModel):
    """Area represents a named zone within a location with polygon boundary"""

    id: str
    name: str
    description: Optional[str] = None
    boundary: List[Coordinate]  # Polygon points
    fill_color: Optional[str] = Field(default=None, alias="fillColor")
    border_color: Optional[str] = Field(default=None, alias="borderColor")
    opacity: Optional[float] = None  # 0.0 to 1.0
    type: Optional[str] = None  # perimeter, patrol_zone, checkpoint, etc.
    priority: Optional[str] = None  # high, medium, low
    active: bool
    created_at: datetime = Field(alias="createdAt")
    updated_at: datetime = Field(alias="updatedAt")

    class Config:
        populate_by_name = True


class Location(BaseModel):
    """Location represents a geographic location with center coordinates and multiple areas"""

    id: str = Field(alias="id")
    name: str
    description: Optional[str] = None

    # Center coordinates
    latitude: float
    longitude: float

    # Areas within this location
    areas: List[Area] = Field(default_factory=list)

    # Visual settings
    color: Optional[str] = None
    icon: Optional[str] = None

    # Metadata
    use_case: Optional[str] = Field(default=None, alias="useCase")
    tags: List[str] = Field(default_factory=list)

    # Status
    active: bool

    # Audit
    created_by: Optional[str] = Field(default=None, alias="createdBy")
    created_at: datetime = Field(alias="createdAt")
    updated_by: Optional[str] = Field(default=None, alias="updatedBy")
    updated_at: datetime = Field(alias="updatedAt")

    class Config:
        populate_by_name = True


class CreateLocationRequest(BaseModel):
    """Request to create a new location"""

    name: str
    description: Optional[str] = None
    latitude: float
    longitude: float
    color: Optional[str] = None
    icon: Optional[str] = None
    use_case: Optional[str] = Field(default=None, alias="useCase")
    tags: Optional[List[str]] = None
    active: Optional[bool] = None

    class Config:
        populate_by_name = True


class GetLocationsRequest(BaseModel):
    """Request to get locations"""

    use_case: Optional[str] = Field(default=None, alias="useCase")
    active: Optional[bool] = None

    class Config:
        populate_by_name = True
